apps/grid/reframe_grid.py

```diff
@@ -0,0 +1,56 @@
+"""
+ReFrame benchmark for GIRD: https://github.com/paboyle/Grid
+
+Assumes a GRID install is present on the system. Developed from
+https://github.com/DiRAC-HPC/es-benchmarking-public
+
+WARNING:
+The code is Hybrid OpenMP + MPI with NUMA socket aware optimisations.
+The relevant options can make big changes to delivered performance.
+"""
+import reframe as rfm
+import reframe.utility.sanity as sn
+
+
+@rfm.simple_test
+class GridBenchmark(rfm.RunOnlyRegressionTest):
+
+    valid_systems = ['*']
+    valid_prog_environs = ['intel20-mpi']
+
+    num_tasks = 1
+    num_tasks_per_node = 1
+    num_cpus_per_task = 20
+
+    executable = 'Benchmark_ITT'
+    executable_opts = ['--shm 1024 --shm-hugetlb']
+    time_limit = '20m'
+    variables = {
+        'OMP_NUM_THREADS': f'{num_cpus_per_task}',
+    }
+    extra_resources = {
+        'mpi': {'num_slots': num_tasks * num_cpus_per_task}
+    }
+
+    reference = {
+        '*': {'Performance': (150000, None, None, 'Mflop/s per node'),
+        }
+    }
+
+    @run_before('sanity')
+    def set_sanity_patterns(self):
+        self.sanity_patterns = sn.assert_found(r'Per Node Summary table',
+                                               self.stdout)
+
+    @run_before('performance')
+    def set_perf_patterns(self):
+        """
+        e.g.
+
+        Grid : Message : 380809 ms :  Comparison point  result: 143382.7 Mflop/s per node
+        """
+
+        self.perf_patterns = {
+            'Performance':
+                sn.extractsingle('.*?result:.+', self.stdout, 0, lambda x: float(x.split()[-4]))
```

Comment:
If anyone knows how to do this with pure regex that would be awesome 😄

Reply:
```suggestion
                sn.extractsingle('.+result: (\S+).+', self.stdout, 0, lambda x: float(x.split()[-4]))
```
tried it on [regex101](https://regex101.com/) and it seems like ^ is needed(?)

Reply:
I can't tell the difference apart from the fact this starts with `.+`, but we aren't really interested in the fact the line doesn't start with `result` (and we don't need to match the start of the string in the other benchmarks we have).

Reply:
It appears that above I simply put an extra `s` in `Mflops/s` (no idea from where since I thought I copied the string from the code example
```suggestion
                sn.extractsingle('result: (\S+) Mflop/s per node', self.stdout, 1, float)
```

Side note, when I want to try out quickly performance benchmarks like this I create a dummy test which prints out what I want:
```py
import reframe as rfm
import reframe.utility.sanity as sn

@rfm.simple_test
class Test(rfm.RegressionTest):
    valid_systems = ['*']
    valid_prog_environs = ['*']
    build_system = 'Spack'
    executable = 'echo'
    executable_opts = ['Grid : Message : 380809 ms :  Comparison point  result: 143382.7 Mflop/s per node']
    reference = {
        'csd3:icelake': {
            'Performance': (228286, None, None, 'Mflop/s per node')
        },
        '*': {
            'Performance': (150000, None, None, 'Mflop/s per node'),
        }
    }

    @run_before('sanity')
    def set_sanity_patterns(self):
        self.sanity_patterns = sn.assert_found(r'result: (\S+) Mflop/s per node', self.stdout)

    @run_before('performance')
    def set_perf_patterns(self):
        """
        e.g.
        Grid : Message : 380809 ms :  Comparison point  result: 143382.7 Mflop/s per node
        """

        self.perf_patterns = {
            'Performance':
            sn.extractsingle(r'result: (\S+) Mflop/s per node', self.stdout, 1, float)
        }
```
```console
% reframe -c test.py -r --performance-report 
[ReFrame Setup]
  version:           3.12.0-dev.0+cec5eee8
  command:           '/home/mose/repo/reframe/bin/reframe -c test.py -r --performance-report'
  launched by:       mose@dream
  working directory: '/tmp'
  settings file:     '<builtin>'
  check search path: '/tmp/test.py'
  stage directory:   '/tmp/stage'
  output directory:  '/tmp/output'

[==========] Running 1 check(s)
[==========] Started on Fri May  6 15:50:23 2022 

[----------] start processing checks
[ RUN      ] Test @generic:default+builtin
[       OK ] (1/1) Test @generic:default+builtin
[----------] all spawned checks have finished

[  PASSED  ] Ran 1/1 test case(s) from 1 check(s) (0 failure(s), 0 skipped)
[==========] Finished on Fri May  6 15:50:25 2022 
==============================================================================
PERFORMANCE REPORT
------------------------------------------------------------------------------
Test
- generic:default
   - builtin
      * num_tasks: 1
      * Performance: 143382.7 Mflop/s per node
------------------------------------------------------------------------------
Run report saved in '/home/mose/.reframe/reports/run-report.json'
Log file(s) saved in '/tmp/rfm-l_cqysn_.log'
```